### ISSUE TITLE
Run tests concurrently (internal)

### ISSUE THREAD

**Issue by gar1t** [guildai/guildai#393]
Non-uat tests (i.e. those immediately under `guild/tests`) are designed to run independently of one another (with one exception, see below). We should be able to run these tests concurrently and dramatically improve testing performance (speedup should be close to linear).

An issue we'll run into is that later tests use a pattern of changing into a project and running Guild commands in the context of that project directory. This introduces an obvious locking problem as tests are run in parallel.

**Comment by gar1t** [comment]
The approaches we can take here:

0. Do nothing - continue to run tests in series
1. Stop using any pattern that relies on the current working directory
2. Run tests in a separate OS process
3. Explicitly lock the cwd

Option 3 seems to me the least desirable. It requires lock-awareness from tests and can become the source of deadlocks, adding further complexity to the test facility (timeouts, deadlock checks, etc.)

Option 1 is arguably the correct/pure approach, particularly for unit tests, which should not have system-level side effects. Realistically, Guild tests will have system level side effects so I don't think purity should be a goal in this case.

Option 2 is I think the most viable. It may be a trivial matter to configure a threaded test runner to use Python's [multiprocessing](https://docs.python.org/3/library/multiprocessing.html) facility to isolate each test.

**Comment by gar1t** [comment]
Another issue here is how to display progress. Guild's current linear approach is straight forward - print each test as it's run (in lexicographic order) and its results.

I would prefer to maintain the sorted list, regardless of the order the tests are run in.

Some approaches:

0. Do nothing - continue to run tests in series
1. Run tests in order using queues (one queue per execution thread), maintain results in the background, and only display a result in the corrected sort order when it is available
2. Use curses to maintain status for running tests, maintaining the correct sort order

Option 1 is the least invasive to user experience and to the code. It suffers from not showing what's running (only the 'current' test based on sort order is shown).

Option 2 is more complex but shows real status. We might consider the [blessings](https://github.com/erikrose/blessings) library to maintain status - and also because of its name.

I'm inclined to go with Option 1 given the low impact and directness of the solution.
